src/freenet/client/async/SplitFileInserterStorage.java

```diff
@@ -1760,4 +1762,16 @@ public void clearCooldown() {
         this.callback.clearCooldown();
     }
 
+    /** @return -1 if the insert has finished, 0 if has blocks to send, otherwise Long.MAX_VALUE. */
+    public long getWakeupTime(ClientContext context, long now) {
+        // LOCKING: hasFinished() uses (this), separate from cooldownLock.
+        // It is safe to use both here (on the request selection thread), one after the other.
+        if (hasFinished()) 
+            return -1;
+        if (noBlocksToSend())
+            return Long.MAX_VALUE;
```

Comment:
It looks like this introduces a race condition. Consider the following execution:

Thread1: hasFinishes() → false
Thread2: arbitrary sequence of operations that finish the insert (possible, since no locks are taken)
Thread1: noBlocksToSend() → true; returns `Long.MAX_VALUE` where `-1` is expected, since the request is now finished

A `synchronized(coolDownBlock) { synchronized(this) { ... }}` surrounding this entire function may avoid this. `chooseBlock()` also takes locks in this order.

Note: I'm not sure whether this race condition is actually harmful, I'm merely pointing out that it exists.